src/com/ms/silverking/io/IORelay.java

```diff
@@ -17,168 +15,168 @@
+  }
 
-    public static void appendFromFile(File file, OutputStream out) throws IOException {
-        append(new FileInputStream(file), out);
-    }
-    
-    public static void append(InputStream in, OutputStream out) throws IOException {
-        new IORelay(in, out, false).startAndWait();
-    }
-    
-    public void start() {
-        running = true;
-        ThreadUtil.newDaemonThread(this, "IORelay").start();
-    }
-    
-    public void stop() {
-        running = false;
-    }
-    
-    public void stopAndDrain() {
-        drain = true;
-        waitFor();
-    }    
-    
-    public void waitFor() {
-        synchronized (this) {
-            while (running) {
-                try {
-                    this.wait();
-                } catch (InterruptedException ie) {
-                }
-            }
+  public IORelay(InputStream in, OutputStream out, int bufferSize) {
+    this(in, out, bufferSize, true);
+  }
+
+  public IORelay(InputStream in, OutputStream out) {
+    this(in, out, DEF_BUFFER_SIZE);
+  }
+
+  public IORelay(InputStream in, OutputStream out, boolean close) {
+    this(in, out, DEF_BUFFER_SIZE, close);
+  }
+
+  public static IORelay fileRelay(InputStream in, String fileName) throws IOException {
+    return new IORelay(in, new FileOutputStream(fileName));
+  }
+
+  public static IORelay fileRelay(InputStream in, File file) throws IOException {
+    return new IORelay(in, new FileOutputStream(file));
+  }
+
+  public static void streamToFile(InputStream in, String fileName) throws IOException {
+    fileRelay(in, fileName).startAndWait();
+  }
+
+  public static void streamToFileInBackground(InputStream in, File file) throws IOException {
+    fileRelay(in, file).start();
+  }
+
+  public static void streamToFile(InputStream in, File file) throws IOException {
+    fileRelay(in, file).startAndWait();
+  }
+
+  public static void streamFromFile(String fileName, OutputStream out) throws IOException {
+    streamFromFile(new File(fileName), out);
+  }
+
+  public static void streamFromFile(File file, OutputStream out) throws IOException {
+    stream(new FileInputStream(file), out);
+  }
+
+  public static void stream(InputStream in, OutputStream out) throws IOException {
+    new IORelay(in, out).startAndWait();
+  }
+
+  public static void appendFromFile(File file, OutputStream out) throws IOException {
+    append(new FileInputStream(file), out);
+  }
+
+  public static void append(InputStream in, OutputStream out) throws IOException {
+    new IORelay(in, out, false).startAndWait();
+  }
+
+  public void start() {
+    running = true;
+    ThreadUtil.newDaemonThread(this, "IORelay").start();
+  }
+
+  public void stop() {
+    running = false;
+  }
+
+  public void stopAndDrain() {
+    drain = true;
+    waitFor();
+  }
+
+  public void waitFor() {
+    synchronized (this) {
+      while (running) {
+        try {
+          this.wait();
+        } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)